ibis/pandas/core.py

```diff
@@ -115,311 +116,266 @@
+    # nodes. We have to filter out nodes for cases--such as zero argument
+    # UDFs--that do not have any dependencies yet still need to be evaluated.
+    full_scope = toolz.merge(
+        scope if scope is not None else {},
+        {
+            key: key
+            for key, parents in dependencies.items()
+            if not parents and not isinstance(key, ops.Node)
+        },
+        params,
+    )
 
+    # Call pre_execute, to allow clients to intercept the expression before
+    # computing anything *and* before associating leaf nodes with data. This
+    # allows clients to provide their own data for each leaf.
+    if not clients:
+        clients = list(find_backends(expr))
 
-def execute_and_reset(
-    expr, params=None, scope=None, aggcontext=None, **kwargs
-):
-    """Execute an expression against data that are bound to it. If no data
-    are bound, raise an Exception.
+    if aggcontext is None:
+        aggcontext = agg_ctx.Summarize()
 
-    Notes
-    -----
-    The difference between this function and :func:`~ibis.pandas.core.execute`
-    is that this function resets the index of the result, if the result has
-    an index.
+    # give backends a chance to inject scope if needed
+    execute_first_scope = execute_first(
+        expr.op(), *clients, scope=full_scope, aggcontext=aggcontext, **kwargs
+    )
+    full_scope = toolz.merge(full_scope, execute_first_scope)
 
-    Parameters
-    ----------
-    expr : ibis.expr.types.Expr
-        The expression to execute
-    params : Mapping[ibis.expr.types.Expr, object]
-        The data that an unbound parameter in `expr` maps to
-    scope : Mapping[ibis.expr.operations.Node, object]
-        Additional scope, mapping ibis operations to data
-    aggcontext : Optional[ibis.pandas.aggcontext.AggregationContext]
-        An object indicating how to compute aggregations. For example,
-        a rolling mean needs to be computed differently than the mean of a
-        column.
-    kwargs : Dict[str, object]
-        Additional arguments that can potentially be used by individual node
-        execution
+    nodes = [node for node in toposorted if node not in full_scope]
 
-    Returns
-    -------
-    result : Union[
-        pandas.Series, pandas.DataFrame, ibis.pandas.core.simple_types
-    ]
-
-    Raises
-    ------
-    ValueError
-        * If no data are bound to the input expression
-    """
-    result = execute(
-        expr, params=params, scope=scope, aggcontext=aggcontext, **kwargs
-    )
-    if isinstance(result, pd.DataFrame):
-        schema = expr.schema()
-        df = result.reset_index()
-        return df.loc[:, schema.names]
-    elif isinstance(result, pd.Series):
-        return result.reset_index(drop=True)
+    # compute the nodes that are not currently in scope
+    for node in nodes:
+        # allow clients to pre compute nodes as they like
+        pre_executed_scope = pre_execute(
```

Comment:
we should talk about this a bit. based on the naming, i would expect `pre_execute` to happen before `execute_first`. and i think executing `pre_execute` in a loop over each node is a little different than how it happens now. we may not have good test cases that exercise the nuances of `pre_execute` here.